README.md
[![Build Status](https://travis-ci.org/arve0/fijibin.svg)](https://travis-ci.org/arve0/fijibin)
# fijibin
This software will download the latest *Life-Line* version of
[Fiji](http://fiji.sc/), make the correct cross-platform binary available and
provide a macro submodule which makes automation of Fiji trivial from python.

**Video demo**

[![](http://img.youtube.com/vi/v0q88SisBtw/0.jpg)](http://youtu.be/v0q88SisBtw)

Current Fiji Life-Line version in this package is **2014 November 25**.

If you experience any trouble with this module, please
[submit an issue](https://github.com/arve0/fijibin/issues/new) or send a
pull request on github.

## Install
```
pip install fijibin
```

## Usage
```python
>>> import fijibin
>>> fijibin.BIN
'/Users/arve/.bin/Fiji.app/Contents/MacOS/ImageJ-macosx'
>>> fijibin.FIJI_VERSION
'20141125'
```

`fijibin.BIN` will point to linux, mac or windows version, depending on the
operating system detected via [platform](https://docs.python.org/3.4/library/platform.html).

### Macros
```python
import fijibin.macro
macro.run(macro_string_or_list_of_strings)
```

### Refetch binary
```python
>>> from fijibin.fetch import fetch
>>> fetch()
```

See more in the [API reference](http://fijibin.readthedocs.org/).

## Development
Install dependencies and link development version of fijibin to pip:
```bash
git clone https://github.com/arve0/fijibin
cd fijibin
pip install -r requirements.txt
```

### run test
```bash
pip install tox
tox
```

### extra output, jump into pdb upon error
```bash
DEBUG=fijibin tox -- --pdb -s
```

### build api reference
```bash
pip install -r docs/requirements.txt
make docs
```


changelog.md
# 0.3.0
- fix python2.7 os.fdopen error
- add test
- add travis testing

# 0.2.1
- include VERSION in source package

# 0.2.0
- download fiji on import instead of in setup.py (which never runs for
  wheel installs)

# 0.1.0
- use pydebug for debugging messages
- added API reference docs
- fix JRE on windows
- fix temp file permission trouble on windows

# 0.0.5
- macro.run: return existing output_files even if macro empty

# 0.0.4
- macro.stitch: support for 16 bit images

# 0.0.3
- add fijibin.macro

# 0.0.2
- use 64-bit Fiji if operating system is 64-bit

# 0.0.1
- use `BIN` as variable name
- chmod +x for bin_names after fetch
- VERSION -> FIJI_VERSION
- use Fiji version name
  - Fiji.app -> Fiji-20141125.app

# 2014.11.25 (removed from pypi)
- initial version


tests/test_stitch.py
"Test stitching of Leica OME.TIF experiment."

import pytest
from py import path
from fijibin import macro


@pytest.fixture
def images(tmpdir):
    "'images' in tmpdir. Returns py.path object."
    e = path.local(__file__).dirpath().join('images')
    e.copy(tmpdir.mkdir('experiment'))
    return tmpdir.join('experiment')


def test_stitch(tmpdir, images):
    "It should stitch images without error."
    filenames = 'x{x}y{y}.png'
    outfile = tmpdir.join('stitched.png')
    m = macro.stitch(images.strpath, filenames, 1, 2, outfile.strpath)
    macro.run(m)

    # output file should exists
    assert outfile.check()


fijibin/fetch.py
#!/usr/bin/env python
import sys, os, stat, fijibin
from io import BytesIO
from zipfile import ZipFile
try:
    from zipfile import BadZipZile
except ImportError:
    from zipfile import BadZipfile
    BadZipFile = BadZipfile


def fetch():
    "Fetch and extract latest Life-Line version of Fiji is just ImageJ to `~/.bin`."
    if os.path.isdir(fijibin.FIJI_ROOT):
        return

    try:
        # python 2
        from urllib2 import urlopen, HTTPError, URLError
    except ImportError:
        # python 3
        from urllib.request import urlopen, HTTPError, URLError

    url = fijibin.URL

    try:
        print('Getting fiji from %s' % url)
        req = urlopen(url)
        try:
            size = int(req.info()['content-length'])
        except AttributeError:
            size = -1

        chunk = 512*1024
        fp = BytesIO()
        i = 0
        while 1:
            data = req.read(chunk)
            if not data:
                break
            fp.write(data)
            if size > 0:
                percent = fp.tell() // (size/100)
                msg = 'Downloaded %d percent      \r' % percent
            else:
                msg = 'Downloaded %d bytes\r' % fp.tell()
            sys.stdout.write(msg)
    except (HTTPError, URLError) as e:
        print('Error getting fiji: {}'.format(e))
        sys.exit(1)

    try:
        print('\nExtracting zip')
        z = ZipFile(fp)
        z.extractall(fijibin.BIN_FOLDER)
        # move to Fiji-VERSION.app to easily check if it exists (upon fijibin upgrade)
        os.rename(fijibin.EXTRACT_FOLDER, fijibin.FIJI_ROOT)
    except (BadZipFile, IOError) as e:
        print('Error extracting zip: {}'.format(e))
        sys.exit(1)

    for path in fijibin.BIN_NAMES.values():
        st = os.stat(path)
        os.chmod(path, st.st_mode | stat.S_IEXEC)


fijibin/macro.py
# encoding: utf-8
"""
Run Fiji is just ImageJ macros headless with python.
"""
import pydebug, subprocess, os, re
from tempfile import mkstemp
import fijibin

# debug with DEBUG=fijibin python script.py
debug = pydebug.debug('fijibin')


##
# Running macros
##
def run(macro, output_files=[], force_close=True):
    """
    Runs Fiji with the suplied macro. Output of Fiji can be viewed by
    setting environment variable `DEBUG=fijibin`.

    Parameters
    ----------
    macro : string or list of strings
        IJM-macro(s) to run. If list of strings, it will be joined with
        a space, so all statements should end with ``;``.
    output_files : list
        Files to check if exists after macro has been run. Files specified that
        do not exist after macro is done will print a warning message.
    force_close : bool
        Will add ``eval("script", "System.exit(42);");`` to end of macro. Exit
        code 42 is used to overcome that errors in macro efficiently will exit
        Fiji with error code 0. In other words, if this line in the macro is
        reached, the macro has most probably finished without errors. This
        is the default behaviour.

        One should also note that Fiji doesn't terminate right away if
        ``System.exit()`` is left out, and it may take several minutes for
        Fiji to close.

    Returns
    -------
    int
        Files from output_files which exists after running macro.
    """
    if type(macro) == list:
        macro = ' '.join(macro)
    if len(macro) == 0:
        print('fijibin.macro.run got empty macro, not starting fiji')
        return _exists(output_files)
    if force_close:
        # make sure fiji halts immediately when done
        # hack: use error code 42 to check if macro has run sucessfully
        macro = macro + 'eval("script", "System.exit(42);");'

    # escape backslashes (windows file names)
    #                 not \ \  not \      g1 \\ g2
    macro = re.sub(r"([^\\])\\([^\\])", r"\1\\\\\2", macro)

    debug('macro {}'.format(macro))

    # avoid verbose output of Fiji when DEBUG environment variable set
    env = os.environ.copy()
    debugging = False
    if 'DEBUG' in env:
        if env['DEBUG'] == 'fijibin' or env['DEBUG'] == '*':
            debugging = True
        del env['DEBUG']

    fptr, temp_filename = mkstemp(suffix='.ijm')
    m = os.fdopen(fptr, 'w')
    m.write(macro)
    m.flush() # make sure macro is written before running Fiji
    m.close()

    cmd = [fijibin.BIN, '--headless', '-macro', temp_filename]
    proc = subprocess.Popen(cmd, stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE, env=env)
    out, err = proc.communicate()

    for line in out.decode('latin1', errors='ignore').splitlines():
        debug('stdout:' + line)
    for line in err.decode('latin1', errors='ignore').splitlines():
        debug('stderr:' + line)

    if force_close and proc.returncode != 42:
        print('fijibin ERROR: Fiji did not successfully ' +
              'run macro {}'.format(temp_filename))
        if not debugging:
            print('fijibin Try running script with ' +
                  '`DEBUG=fijibin python your_script.py`')
    else:
        # only delete if everything is ok
        os.remove(temp_filename)


    # return output_files which exists
    return _exists(output_files)

##
# Collection of macros
##
def stitch(folder, filenames, x_size, y_size, output_filename,
                 x_start=0, y_start=0, overlap=10):
    """
    Creates a Fiji Grid/Collection stitching macro. Parameters are the same as
    in the plugin and are described in further detail here:
    http://fiji.sc/Image_Stitching#Grid.2FCollection_Stitching.

    **Default stitch parameters:**

    * Filename defined positions
    * Compute overlap
    * Subpixel accurancy
    * Save computation time (but use more RAM)
    * Fusion method: Linear blending
    * Regression threshold: 0.30
    * Max/avg displacement threshold: 2.50
    * Absolute displacement threshold: 3.50


    Parameters
    ----------
    folder : string
        Path to folder with images or folders with images.
        Example: */path/to/slide--S00/chamber--U01--V02/*
    filenames : string
        Filenames of images.
        Example: *field-X{xx}-Y{yy}/image-X{xx}-Y{yy}.ome.tif*
    x_size : int
        Size of grid, number of images in x direction.
    y_size : int
        Size of grid, number of images in y direction.
    output_filename : string
        Where to store fused image. Should be `.png`.
    x_start : int
        Which x position grid start with.
    y_start : int
        Which y position grid start with.
    overlap : number
        Tile overlap in percent. Fiji will find the optimal overlap, but a
        precise overlap assumption will decrase computation time.

    Returns
    -------
    string
        IJM-macro.
    """

    macro = []
    macro.append('run("Grid/Collection stitching",')
    macro.append('"type=[Filename defined position]')
    macro.append('order=[Defined by filename         ]')
    macro.append('grid_size_x={}'.format(x_size))
    macro.append('grid_size_y={}'.format(y_size))
    macro.append('tile_overlap={}'.format(overlap))
    macro.append('first_file_index_x={}'.format(x_start))
    macro.append('first_file_index_y={}'.format(y_start))
    macro.append('directory=[{}]'.format(folder))
    macro.append('file_names=[{}]'.format(filenames))
    macro.append('output_textfile_name=TileConfiguration.txt')
    macro.append('fusion_method=[Linear Blending]')
    macro.append('regression_threshold=0.20')
    macro.append('max/avg_displacement_threshold=2.50')
    macro.append('absolute_displacement_threshold=3.50')
    macro.append('compute_overlap')
    macro.append('subpixel_accuracy')
    macro.append('computation_parameters=[Save computation time (but use more RAM)]')
    # use display, such that we can specify output filename
    # this is 'Fused and display' for previous stitching version!!
    macro.append('image_output=[Fuse and display]");')
    # save to png
    macro.append('selectWindow("Fused");')
    macro.append('saveAs("PNG", "{}");'.format(output_filename))
    macro.append('close();')

    return ' '.join(macro)


##
# Helper functions
##
def _exists(filenames):
    """Check if every filename exists. If not, print an error
    message and remove the item from the list.

    Parameters
    ----------
    filenames : list
        List of filenames to check for existence.

    Returns
    -------
    list
        Filtered list of filenames that exists.
    """
    exists = []
    for filename in filenames:
        if os.path.isfile(filename):
            exists.append(filename)
        else:
            print('fijibin ERROR missing output file {}'.format(filename))

    return exists


fijibin/__init__.py
"""
Latest Life-Line version of fiji for easy inclusion in Python projects.
"""
import sys, os, platform, stat, shutil
from io import BytesIO
from zipfile import ZipFile
from os.path import join, dirname
from . import macro
try:
    from zipfile import BadZipZile
except ImportError:
    from zipfile import BadZipfile
    BadZipFile = BadZipfile

__all__ = ['fetch', 'macro']


##
# CONSTANTS
##

VERSION = open(join(dirname(__file__), 'VERSION')).read().strip()
FIJI_VERSION = '20141125'
BASE_URL = 'http://fiji.sc/downloads/Life-Line/fiji-'
END_URL = '-' + FIJI_VERSION + '.zip'

HOME = os.path.expanduser('~')
BIN_FOLDER = join(HOME, '.bin')
EXTRACT_FOLDER = join(BIN_FOLDER, 'Fiji.app')
FIJI_ROOT = join(BIN_FOLDER, 'Fiji-' + FIJI_VERSION + '.app')
BIN_NAMES = {
    'Darwin64bit': join(FIJI_ROOT, 'Contents/MacOS/ImageJ-macosx'),
    'Linux32bit': join(FIJI_ROOT, 'ImageJ-linux32'),
    'Linux64bit': join(FIJI_ROOT, 'ImageJ-linux64'),
    'Windows32bit': join(FIJI_ROOT, 'ImageJ-win32.exe'),
    'Windows64bit': join(FIJI_ROOT, 'ImageJ-win64.exe')
}

_unix_url = BASE_URL + 'nojre' + END_URL
URLS = {
    'Darwin64bit': _unix_url,
    'Linux32bit': _unix_url,
    'Linux64bit': _unix_url,
    'Windows32bit': BASE_URL + 'win32' + END_URL,
    'Windows64bit': BASE_URL + 'win64'  + END_URL
}

SYSTEM = platform.system() + platform.architecture()[0]
BIN = BIN_NAMES[SYSTEM]
URL = URLS[SYSTEM]


##
# FETCH BINARY
##
def fetch(force=False):
    """Fetch and extract latest Life-Line version of Fiji is just ImageJ
    to *~/.bin*.

    Parameters
    ----------
    force : bool
        Force overwrite of existing Fiji in *~/.bin*.

    """
    try:
        # python 2
        from urllib2 import urlopen, HTTPError, URLError
    except ImportError:
        # python 3
        from urllib.request import urlopen, HTTPError, URLError

    if os.path.isdir(FIJI_ROOT) and not force:
        return
    elif not os.path.isdir(FIJI_ROOT):
        print('Fiji missing in %s' % FIJI_ROOT)

    if force:
        print('Deleting %s' % FIJI_ROOT)
        shutil.rmtree(FIJI_ROOT, ignore_errors=True)

    print('Downloading fiji from %s' % URL)
    try:
        req = urlopen(URL)
        try:
            size = int(req.info()['content-length'])
        except AttributeError:
            size = -1

        chunk = 512*1024
        fp = BytesIO()
        i = 0
        while 1:
            data = req.read(chunk)
            if not data:
                break
            fp.write(data)
            if size > 0:
                percent = fp.tell() // (size/100)
                msg = 'Downloaded %d percent      \r' % percent
            else:
                msg = 'Downloaded %d bytes\r' % fp.tell()
            sys.stdout.write(msg)
    except (HTTPError, URLError) as e:
        print('Error getting fiji: {}'.format(e))
        sys.exit(1)

    try:
        print('\nExtracting zip')
        z = ZipFile(fp)
        z.extractall(BIN_FOLDER)
        # move to Fiji-VERSION.app to easily check if it exists (upon fijibin upgrade)
        os.rename(EXTRACT_FOLDER, FIJI_ROOT)
    except (BadZipFile, IOError) as e:
        print('Error extracting zip: {}'.format(e))
        sys.exit(1)

    for path in BIN_NAMES.values():
        st = os.stat(path)
        os.chmod(path, st.st_mode | stat.S_IEXEC)

##
# Download upon import if Fiji is missing
##
fetch()
